src/main/java/org/elasticsearch/search/aggregations/reducers/BucketHelpers.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.reducers;
+
+import org.elasticsearch.ElasticsearchIllegalStateException;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.SearchParseException;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.InvalidAggregationPathException;
+import org.elasticsearch.search.aggregations.metrics.InternalNumericMetricsAggregation;
+import org.elasticsearch.search.aggregations.reducers.derivative.DerivativeParser;
+import org.elasticsearch.search.aggregations.support.AggregationPath;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A set of static helpers to simplify working with aggregation buckets, in particular
+ * providing utilities that help reducers.
+ */
+public class BucketHelpers {
+
+    /**
+     * A gap policy determines how "holes" in a set of buckets should be handled.  For example,
+     * a date_histogram might have empty buckets due to no data existing for that time interval.
+     * This can cause problems for operations like a derivative, which relies on a continuous
+     * function.
+     *
+     * "insert_zeros": empty buckets will be filled with zeros for all metrics
+     * "ignore": empty buckets will simply be ignored
+     */
+    public static enum GapPolicy {
+        INSERT_ZEROS((byte) 0, "insert_zeros"), IGNORE((byte) 1, "ignore");
```

Comment:
I am thinking about renaming IGNORE to SKIP as this seems to better describe what it is doing. The missing values are not really ignored, it's more that we skip over the missing values. thoughts?